File: android_mobile/src/main/java/com/alexstyl/specialdates/contact/AndroidContactsProviderSource.java
package com.alexstyl.specialdates.contact;

import java.util.List;

class AndroidContactsProviderSource implements ContactsProviderSource {

    private final ContactCache<Contact> cache;
    private final AndroidContactFactory factory;

    AndroidContactsProviderSource(ContactCache<Contact> contactCache, AndroidContactFactory factory) {
        this.cache = contactCache;
        this.factory = factory;
    }

    @Override
    public Contact getOrCreateContact(long contactID) throws ContactNotFoundException {
        Contact deviceContact = cache.getContact(contactID);
        if (deviceContact == null) {
            deviceContact = factory.createContactWithId(contactID);
            cache.addContact(deviceContact);
        }
        return deviceContact;
    }

    @Override
    public List<Contact> getAllContacts() {
        List<Contact> allContacts = factory.getAllContacts();
        cache.evictAll();
        for (Contact allContact : allContacts) {
            cache.addContact(allContact);
        }
        return allContacts;
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/contact/AndroidContactsProviderSource.kt
package com.alexstyl.specialdates.contact

class AndroidContactsProviderSource(private val cache: ContactCache,
                                    private val factory: AndroidContactFactory
) : ContactsProviderSource {

    // TODO evict if issue

    @Throws(ContactNotFoundException::class)
    override fun getOrCreateContact(contactID: Long): Contact {
        var deviceContact = cache.getContact(contactID)
        if (deviceContact == null) {
            deviceContact = factory.createContactWithId(contactID)
            cache.addContact(deviceContact)
        }
        return deviceContact
    }

    override val allContacts: List<Contact>
        get() {
            val allContacts = factory.getAllContacts()
            cache.evictAll()
            cache.addContacts(allContacts)
            return allContacts
        }

    override fun queryContacts(contactIds: List<Long>): List<Contact> {
        val contacts = factory.queryContacts(contactIds)
        cache.addContacts(contacts)
        return contacts
    }
}
